### PR TITLE
Handle case where state is a tensor

### PR DESCRIPTION
**Pull request by gokulavasan**
Fixes the case where tensor is returned as the state instead of a dict. This was failing because the tensor object was being used as a bool check for the walrus operator and that failed. Instead just do regular fetch and not None comparison.

Added unit tests in all the layers to cover this

Fixes #{1271}

### Changes

-
-

### CODE REVIEW

**Comment by gokulavasan** [comment]
Fix is in this PR - https://github.com/pytorch/data/pull/1272

File: test/stateful_dataloader/test_state_dict.py [hunk 1]
```diff
@@ -426,6 +428,32 @@ def load_state_dict(self, state):
         self.epoch = state["epoch"]
 
 
+class StateType(Enum):
+    INTEGER = (1,)
+    STRING = (2,)
+    TENSOR = (3,)
```
Comment (andrewkho) [inline test/stateful_dataloader/test_state_dict.py:434]: I don't really care because this is just a test, but generally for python just use a string, I find this boilerplate just adds friction and doesn't usually help

File: torchdata/stateful_dataloader/incremental_state.py [hunk 2]
```diff
@@ -120,7 +120,8 @@ def __init__(self, initial_worker_state_dict: Optional[Dict[str, Any]]):
         if initial_worker_state_dict:
             self._worker_id = initial_worker_state_dict[_WORKER_ID]
             dataset_state = initial_worker_state_dict.get(_DATASET_STATE, None)
-            if fetcher_state := initial_worker_state_dict.get(_FETCHER_STATE, None):
+            fetcher_state = initial_worker_state_dict.get(_FETCHER_STATE, None)
+            if fetcher_state is not None:
```
Comment (andrewkho) [inline torchdata/stateful_dataloader/incremental_state.py:124]: Why this change?

File: test/stateful_dataloader/test_state_dict.py [hunk 3]
```diff
@@ -1450,6 +1478,23 @@ def test_fast_state_dict_request_skip_steps(self) -> None:
...
+                dl = StatefulDataLoader(
+                    dataset=ds,
+                    num_workers=num_workers,
+                    collate_fn=identity,
+                    multiprocessing_context=("forkserver" if IS_MACOS and num_workers else None),
+                )
+                it = iter(dl)
+                next(it)
+                sd = dl.state_dict()
+                self.assertTrue(sd)
```
Comment (andrewkho) [inline test/stateful_dataloader/test_state_dict.py:1495]: what is this testing?

File: test/stateful_dataloader/test_state_dict.py (hunk 1 above)
Comment (andrewkho) [reply test/stateful_dataloader/test_state_dict.py:434]: Oh I see from summary now

File: torchdata/stateful_dataloader/incremental_state.py (hunk 2 above)
Comment (andrewkho) [reply torchdata/stateful_dataloader/incremental_state.py:124]: Oh I see from summary now

File: test/stateful_dataloader/test_state_dict.py (hunk 1 above)
Comment (andrewkho) [reply test/stateful_dataloader/test_state_dict.py:434]:
```suggestion
```

File: test/stateful_dataloader/test_state_dict.py [hunk 4]
```diff
@@ -426,6 +428,32 @@ def load_state_dict(self, state):
...
+    def load_state_dict(self, state):
+        pass
+
+    def state_dict(self):
+        if self.state_type == StateType.INTEGER:
+            return self.len
+        if self.state_type == StateType.STRING:
+            return str(self.len)
+        if self.state_type == StateType.TENSOR:
+            return torch.rand(self.len)
```
Comment (andrewkho) [inline test/stateful_dataloader/test_state_dict.py:454]:
```suggestion
        return self.len
```

File: test/stateful_dataloader/test_state_dict.py [hunk 5]
```diff
@@ -426,6 +428,32 @@ def load_state_dict(self, state):
...
+class StateType(Enum):
+    INTEGER = (1,)
+    STRING = (2,)
+    TENSOR = (3,)
+
+
+class TensorStateIterableDataset(torch.utils.data.IterableDataset, Stateful):
+    def __init__(self, len, state_type):
+        self.len = len
+        self.state_type = state_type
```
Comment (andrewkho) [inline test/stateful_dataloader/test_state_dict.py:440]:
```suggestion
    def __init__(self, len):
        self.len = len
```

File: test/stateful_dataloader/test_state_dict.py [hunk 6]
```diff
@@ -1450,6 +1478,23 @@ def test_fast_state_dict_request_skip_steps(self) -> None:
         self._run_test(17, 19)
 
 
+class TestNonDictState_shard0(TestCase):
+    def test(self):
+        for state_type in StateType:
```
Comment (andrewkho) [inline test/stateful_dataloader/test_state_dict.py:1483]:
```suggestion
        for ctr in [int, str, torch.rand]:
```

File: test/stateful_dataloader/test_state_dict.py [hunk 7]
```diff
@@ -1450,6 +1478,23 @@ def test_fast_state_dict_request_skip_steps(self) -> None:
...
+            for num_workers in [0, 2]:
+                ds = TensorStateIterableDataset(5, state_type)
```
Comment (andrewkho) [inline test/stateful_dataloader/test_state_dict.py:1485]:
```suggestion
                ds = TensorStateIterableDataset(ctr(5))
```

**Review by andrewkho — APPROVED** [review]
LGTM please simplify the test per instructions before landing and maybe beef up the assert statement a bit
